File: Web/ClientApp/src/app/Ontologies/Deals.ts
import { DealLifeCycleIdentifier, DealStageIdentifier, RestrictedClassifierIdentifier } from "../Deals";
import { LegalEntityIdentifier } from "../LegalEntities";
import { DataAllValuesFrom } from "../Ontology/DataAllValuesFrom";
import { DataComplementOf } from "../Ontology/DataComplementOf";
import { DataOneOf } from "../Ontology/DataOneOf";
import { DisjointClasses } from "../Ontology/DisjointClasses";
import { IClass } from "../Ontology/IClass";
import { INamedIndividual } from "../Ontology/INamedIndividual";
import { IDataPropertyExpression, IObjectPropertyExpression } from "../Ontology/IPropertyExpression";
import { ObjectOneOf } from "../Ontology/ObjectOneOf";
import { ObjectSomeValuesFrom } from "../Ontology/ObjectSomeValuesFrom";
import { Ontology } from "../Ontology/Ontology";
import { DateTime, Decimal } from "../Ontology/Xsd";
import { agreements } from './Agreements';
import { annotations } from './Annotations';
import { commonDomainObjects } from "./CommonDomainObjects";
import { facilityAgreements } from './FacilityAgreements';
import { legalEntities } from "./LegalEntities";
import { lifeCycles } from "./LifeCycles";
import { parties } from "./Parties";
import { roleIndividuals } from "./RoleIndividuals";
import { roles } from "./Roles";

export class Deals extends Ontology
{
    readonly DealType               : IClass;
    readonly Deal                   : IClass;
    readonly Debt                   : IClass;
    readonly Restricted             : IClass;
    readonly Class                  : IDataPropertyExpression;
    readonly Type                   : IObjectPropertyExpression;
    readonly LifeCycle              : IObjectPropertyExpression;
    readonly Classifiers            : IObjectPropertyExpression;
    readonly SponsorsNA             : IDataPropertyExpression;
    readonly LenderParty            : IClass;
    readonly AdvisorParty           : IClass;
    readonly SponsorParty           : IClass;
    readonly BorrowerParty          : IClass;
    readonly Equity                 : IDataPropertyExpression;
    readonly Bank                   : INamedIndividual;
    readonly BankParty              : IClass;
    readonly BankLenderParty        : IClass;
    readonly BankAdvisorParty       : IClass;
    readonly Sponsored              : IClass;
    readonly SponsoredWhenApplicable: IClass;
    readonly NotSponsored           : IClass;
    readonly KeyCounterpartyRole    : IClass;
    readonly KeyCounterparty        : IClass;
    readonly DebtLifeCycle          : INamedIndividual;
    readonly NotRestricted          : INamedIndividual;

    constructor()
    {
        super(
            "Deals",
            commonDomainObjects,
            roles,
            roleIndividuals,
            legalEntities,
            parties,
            lifeCycles,
            annotations,
            agreements,
            facilityAgreements);

        this.DealType = this.DeclareClass("DealType");
        this.DealType.SubClassOf(commonDomainObjects.Classifier);

        this.Deal = this.DeclareClass("Deal");
        this.Deal.SubClassOf(commonDomainObjects.Named);

        this.Class = this.DeclareFunctionalDataProperty("ClassIri");
        this.Classifiers = this.DeclareObjectProperty("Classifiers");
        this.LifeCycle   = this.DeclareFunctionalObjectProperty("LifeCycle");
        this.Type        = this.DeclareFunctionalObjectProperty("Type");
        this.SponsorsNA  = this.DeclareFunctionalDataProperty("SponsorsNA");

        this.Deal.SubClassOf(this.Type.ExactCardinality(1));

        let nonEmptyString = new DataComplementOf(new DataOneOf([""]));

        this.Deal.SubClassOf(commonDomainObjects.Name.MinCardinality(1, nonEmptyString))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);
        this.Deal.SubClassOf(this.DeclareFunctionalObjectProperty("GeographicRegion").MinCardinality(1))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);
        this.Deal.SubClassOf(this.DeclareFunctionalObjectProperty("Currency").MinCardinality(1))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);
        this.Deal.SubClassOf(this.DeclareFunctionalDataProperty("Introducer").MinCardinality(1, nonEmptyString))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);
        this.Deal.SubClassOf(this.DeclareFunctionalDataProperty("TransactionDetails").MinCardinality(1, nonEmptyString))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);
        this.Deal.SubClassOf(this.DeclareFunctionalDataProperty("CurrentStatus").MinCardinality(1, nonEmptyString))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);

        let restrictedClassifier = this.DeclareClass("RestrictedClassifier");
        restrictedClassifier.SubClassOf(commonDomainObjects.Classifier);
        restrictedClassifier.Define(commonDomainObjects.$type.HasValue('Web.Model.RestrictedClassifier, Web'));
        let notRestricted = restrictedClassifier.DeclareNamedIndividual("Not Restricted");
        notRestricted.DataPropertyValue(
            commonDomainObjects.Id,
            RestrictedClassifierIdentifier.No);
        this.Restricted = this.DeclareClass("RestrictedDeal");
        this.Restricted.Define(new ObjectSomeValuesFrom(
            this.Classifiers,
            restrictedClassifier.Intersect(new ObjectOneOf([notRestricted]).Complement())));
        this.Restricted.SubClassOf(this.DeclareFunctionalDataProperty("ProjectName").MinCardinality(1, nonEmptyString))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);

        let exclusivityClassifier = this.DeclareClass("ExclusivityClassifier");
        exclusivityClassifier.SubClassOf(commonDomainObjects.Classifier);
        exclusivityClassifier.Define(commonDomainObjects.$type.HasValue('Web.Model.ExclusivityClassifier, Web'));

        this.Deal.SubClassOf(this.Classifiers.ExactCardinality(1, exclusivityClassifier))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.BusinessScreened)
            .Annotate(annotations.NominalProperty, "Exclusivity");

        let exclusivity = this.DeclareClass("Exclusivity");
        exclusivity.Define(commonDomainObjects.$type.HasValue('Web.Model.Exclusivity, Web'));
        let endDate = this.DeclareFunctionalDataProperty("EndDate");
        endDate.Range(DateTime);
        exclusivity.SubClassOf(endDate.MinCardinality(1))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);


        this.LenderParty   = this.DeclareClass("LenderParty"  );
        this.AdvisorParty  = this.DeclareClass("AdvisorParty" );
        this.BorrowerParty = this.DeclareClass("BorrowerParty");
        this.SponsorParty  = this.DeclareClass("SponsorParty" );

        this.LenderParty.SubClassOf(parties.PartyInRole);
        this.AdvisorParty.SubClassOf(parties.PartyInRole);
        this.BorrowerParty.SubClassOf(parties.PartyInRole);
        this.SponsorParty.SubClassOf(parties.PartyInRole);

        this.LenderParty.Define(parties.Role.HasValue(roleIndividuals.Lender));
        this.AdvisorParty.Define(parties.Role.HasValue(roleIndividuals.Advisor));
        this.BorrowerParty.Define(parties.Role.HasValue(roleIndividuals.Borrower));
        this.SponsorParty.Define(parties.Role.HasValue(roleIndividuals.Sponsor));
        this.Equity = this.DeclareFunctionalDataProperty("Equity");
        this.Equity.Range(Decimal);
        this.SponsorParty.SubClassOf(this.Equity.ExactCardinality(1))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect);

        this.Bank = legalEntities.LegalEntity.DeclareNamedIndividual("Bank");
        this.Bank.DataPropertyValue(commonDomainObjects.Id, LegalEntityIdentifier.Bank);

        this.BankParty        = this.DeclareClass("BankParty");
        this.BankParty.SubClassOf(parties.PartyInRole);
        this.BankParty.Define(parties.Organisation.HasValue(this.Bank));

        this.BankLenderParty  = this.DeclareClass("BankLenderParty");
        this.BankAdvisorParty = this.DeclareClass("BankAdvisorParty");
        this.BankLenderParty.Define(this.LenderParty.Intersect(this.BankParty));
        this.BankAdvisorParty.Define(this.AdvisorParty.Intersect(this.BankParty));

        this.Sponsored = this.DeclareClass("Sponsored");
        this.Sponsored.SubClassOf(this.Deal);
        this.Sponsored.SubClassOf(new ObjectSomeValuesFrom(agreements.Parties, this.SponsorParty))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect)
            .Annotate(annotations.NominalProperty, "Sponsors");
        this.Sponsored.SubClassOf(this.SponsorsNA.ExactCardinality(0));
        this.Sponsored.Annotate(annotations.ComponentBuildAction, "AddSponsors");

        this.SponsoredWhenApplicable = this.DeclareClass("SponsoredWhenApplicable");
        this.SponsoredWhenApplicable.SubClassOf(this.Deal);
        this.SponsoredWhenApplicable.SubClassOf(new ObjectSomeValuesFrom(agreements.Parties, this.SponsorParty).Union(this.SponsorsNA.HasValue(true)))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect)
            .Annotate(annotations.NominalProperty, "Sponsors");
        this.SponsoredWhenApplicable.Annotate(annotations.ComponentBuildAction, "AddSponsors");
        this.SponsoredWhenApplicable.Annotate(annotations.ComponentBuildAction, "AddSponsorsNA");

        this.NotSponsored = this.DeclareClass("NotSponsored");
        this.NotSponsored.SubClassOf(this.Deal);
        this.NotSponsored.SubClassOf(agreements.Parties.ExactCardinality(0, this.SponsorParty));
        this.NotSponsored.SubClassOf(this.SponsorsNA.ExactCardinality(0));

        let sponsoredDeal = this.DeclareClass("SponsoredDeal");
        sponsoredDeal.Define(new ObjectSomeValuesFrom(agreements.Parties, this.SponsorParty));
        sponsoredDeal.SubClassOf(this.Deal);
        sponsoredDeal.SubClassOf(new DataAllValuesFrom(this.DeclareFunctionalDataProperty("TotalSponsorEquity"), new DataOneOf([100])))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect)
            .Annotate(annotations.Error, "MustBe100Percent");

        this.Debt = this.DeclareClass("Debt");
        this.Debt.SubClassOf(this.Deal);
        this.Debt.SubClassOf(agreements.Parties.MinCardinality(1, this.BorrowerParty))
            .Annotate(annotations.RestrictedfromStage, DealStageIdentifier.Prospect)
            .Annotate(annotations.NominalProperty, "Borrowers");
        this.Debt.Annotate(annotations.ComponentBuildAction, "AddDebtTabs");

        this.DebtLifeCycle = lifeCycles.LifeCycle.DeclareNamedIndividual("DebtLifeCycle");
        this.DebtLifeCycle.DataPropertyValue(commonDomainObjects.Id, DealLifeCycleIdentifier.Debt);
        this.Debt.SubClassOf(this.LifeCycle.HasValue(this.DebtLifeCycle));

        this.NotRestricted = this.DeclareNamedIndividual("NotRestricted");
        this.NotRestricted.DataPropertyValue(commonDomainObjects.Id, RestrictedClassifierIdentifier.No);
        this.NotRestricted.DataPropertyValue(commonDomainObjects.$type, 'Web.Model.RestrictedClassifier, Web');

        new DisjointClasses(this, [this.Deal, parties.PartyInRole, commonDomainObjects.Classifier]);
    }
}

export const deals = new Deals();

File: Deals/Deals.cs
using CommonDomainObjects;
using LegalEntities;
using Organisations;
using Ontology;
using Roles;
using System;
using System.Collections.Generic;
using System.Linq;
using Parties;
using People;
using Agreements;

namespace Deals
{
    public static class LegalEntityIdentifier
    {
        public static readonly Guid Mufg = Guid.Empty;
    }

    public class CommonDomainObjects: Ontology.Ontology
    {
        public IClass                  DomainObject { get; protected set; }
        public IDataPropertyExpression Id           { get; protected set; }
        public IClass                  Named        { get; protected set; }
        public IDataPropertyExpression Name         { get; protected set; }
        public IClass                  Classifier   { get; protected set; }

        public CommonDomainObjects() : base("CommonDomainObjects")
        {
            DomainObject = this.DeclareClass<DomainObject<Guid>>();
            Id = DomainObject.DeclareDataProperty<DomainObject<Guid>, Guid>(domainObject => domainObject.Id);
            DomainObject.HasKey(Id);

            Named = this.DeclareClass<Named<Guid>>();
            Named.SubClassOf(DomainObject);
            Name = Named.DeclareDataProperty<Named<Guid>, string>(named => named.Name);

            Classifier = this.DeclareClass<Classifier>();
            Classifier.SubClassOf(Named);
        }
    }

    public class Validation: Ontology.Ontology
    {
        public IAnnotationProperty Restriction     { get; protected set; }
        public IAnnotationProperty SubPropertyName { get; protected set; }
        public IAnnotationProperty RangeValidated  { get; protected set; }

        public Validation() : base("Validation")
        {
            Restriction = new AnnotationProperty(
                this,
                "Restriction");
            SubPropertyName = new AnnotationProperty(
                this,
                "SubPropertyName");
            RangeValidated = new AnnotationProperty(
                this,
                "RangeValidated");
        }
    }

    public class Roles: Ontology.Ontology
    {
        public IClass Role { get; protected set; }

        public Roles(
            CommonDomainObjects commonDomainObjects
            ) : base(
                "Roles",
                commonDomainObjects)
        {
            Role = this.DeclareClass<Role>();
            Role.SubClassOf(commonDomainObjects.Named);
        }
    }

    public class Organisations: Ontology.Ontology
    {
        public IClass Organisation { get; protected set; }

        public Organisations(
            CommonDomainObjects commonDomainObjects
            ) : base(
                "Organisations",
                commonDomainObjects)
        {
            Organisation = this.DeclareClass<Organisation>();
            Organisation.SubClassOf(commonDomainObjects.Named);
        }
    }

    public class LegalEntities: Ontology.Ontology
    {
        public IClass LegalEntity { get; protected set; }

        public LegalEntities(
            Organisations organisations
            ) : base(
                "LegalEntities",
                organisations)
        {
            LegalEntity = this.DeclareClass<LegalEntity>();
            LegalEntity.SubClassOf(organisations.Organisation);
        }
    }

    public class Parties: Ontology.Ontology
    {
        public IClass                    PartyInRole  { get; protected set; }
        public IObjectPropertyExpression Role         { get; protected set; }
        public IObjectPropertyExpression Organisation { get; protected set; }
        public IObjectPropertyExpression Person       { get; protected set; }

        public Parties(
            CommonDomainObjects commonDomainObjects
            ) : base(
                "Parties",
                commonDomainObjects)
        {
            PartyInRole = this.DeclareClass<PartyInRole>();
            PartyInRole.SubClassOf(commonDomainObjects.DomainObject);
            Role         = PartyInRole.DeclareObjectProperty<PartyInRole, Role        >(partyInRole => partyInRole.Role        );
            Organisation = PartyInRole.DeclareObjectProperty<PartyInRole, Organisation>(partyInRole => partyInRole.Organisation);
            Person       = PartyInRole.DeclareObjectProperty<PartyInRole, Person      >(partyInRole => partyInRole.Person      );
        }
    }

    public class RoleIndividuals: Ontology.Ontology
    {
        public readonly INamedIndividual Sponsor;
        public readonly INamedIndividual Borrower;
        public readonly INamedIndividual Lender;
        public readonly INamedIndividual Advisor;

        public RoleIndividuals(
            CommonDomainObjects commonDomainObjects,
            Roles               roles
            ) : base(
                "RoleIndividuals",
                commonDomainObjects,
                roles)
        {
            var role = roles.Role;
            var id = commonDomainObjects.Id;
            Sponsor  = role.DeclareNamedIndividual("Sponsor" );
            Borrower = role.DeclareNamedIndividual("Borrower");
            Lender   = role.DeclareNamedIndividual("Lender"  );
            Advisor  = role.DeclareNamedIndividual("Advisor" );
            Sponsor.Value(id, DealRoleIdentifier.Sponsor);
            Borrower.Value(id, DealRoleIdentifier.Borrower);
            Lender.Value(id, DealRoleIdentifier.Lender);
            Advisor.Value(id, DealRoleIdentifier.Advisor);
        }
    }

    public class Deals: Ontology.Ontology
    {
        public IClass                    DealType            { get; protected set; }
        public IClass                    Deal                { get; protected set; }
        public IClass                    Debt                { get; protected set; }
        public IClass                    Advisory            { get; protected set; }

        public IObjectPropertyExpression Type                { get; protected set; }
        public IObjectPropertyExpression Parties             { get; protected set; }
        public IObjectPropertyExpression Confers             { get; protected set; }
        public IObjectPropertyExpression Classifiers         { get; protected set; }
        public IObjectPropertyExpression Borrowers           { get; protected set; }
        public IObjectPropertyExpression Sponsors            { get; protected set; }
        public IObjectPropertyExpression Exclusivity         { get; protected set; }
        
        public IClass                    LenderParty         { get; protected set; }
        public IClass                    AdvisorParty        { get; protected set; }
        public IClass                    SponsorParty        { get; protected set; }
        public IClass                    BorrowerParty       { get; protected set; }
        public IClass                    MufgParty           { get; protected set; }
        public IClass                    Sponsor             { get; protected set; }
        public IDataPropertyExpression   Equity              { get; protected set; }

        public INamedIndividual          Bank                { get; protected set; }
        public IClass                    BankParty           { get; protected set; }
        public IClass                    BankLenderParty     { get; protected set; }
        public IClass                    BankAdvisorParty    { get; protected set; }
                                         
        public IClass                    KeyCounterpartyRole { get; protected set; }
        public IClass                    KeyCounterparty     { get; protected set; }

        public Deals(
            CommonDomainObjects commonDomainObjects,
            Roles               roles,
            RoleIndividuals     roleIndividuals,
            LegalEntities       legalEntities,
            Parties             parties,
            Validation          validation
            ): base(
                "Deals",
                commonDomainObjects,
                roles,
                roleIndividuals,
                legalEntities,
                parties,
                validation)
        {
            DealType    = this.DeclareClass<DealType>();
            DealType.SubClassOf(commonDomainObjects.Named);
            Deal        = this.DeclareClass<Deal>();
            Deal.SubClassOf(commonDomainObjects.Named);
            Type        = Deal.DeclareObjectProperty<Deal, DealType   >(deal => deal.Type       );
            Parties     = Deal.DeclareObjectProperty<Deal, PartyInRole>(deal => deal.Parties    );
            Classifiers = Deal.DeclareObjectProperty<Deal, Classifier >(deal => deal.Classifiers);
            Confers     = Deal.DeclareObjectProperty<Deal, Commitment >(deal => deal.Confers    );
            Borrowers   = Deal.DeclareObjectProperty<Deal, PartyInRole>(
                "Borrowers",
                deal => deal.Parties.Where(dealParty => dealParty.Role.Id == DealRoleIdentifier.Borrower));
            Sponsors    = Deal.DeclareObjectProperty<Deal, Sponsor   >(
                "Sponsors",
                deal => deal.Parties.Where(dealParty => dealParty.Role.Id == DealRoleIdentifier.Sponsor).Cast<Sponsor>());
            this.Exclusivity = Deal.DeclareObjectProperty<Deal, ExclusivityClassifier>(
                "Exclusivity",
                deal => deal.Classifiers.OfType<ExclusivityClassifier>().FirstOrDefault());

            Deal.SubClassOf(
                new DataSomeValuesFrom(
                    commonDomainObjects.Name,
                    new DataComplementOf(new DataOneOf(string.Empty))))
                .Annotate(
                    validation.Restriction,
                    0);
            Deal.SubClassOf(this.Exclusivity.ExactCardinality(1))
                .Annotate(
                    validation.Restriction,
                    0);

            LenderParty   = this.DeclareClass("LenderParty"  );
            AdvisorParty  = this.DeclareClass("AdvisorParty" );
            BorrowerParty = this.DeclareClass("BorrowerParty");
            SponsorParty  = this.DeclareClass("SponsorParty" );

            LenderParty.SubClassOf(parties.PartyInRole);
            AdvisorParty.SubClassOf(parties.PartyInRole);
            BorrowerParty.SubClassOf(parties.PartyInRole);
            SponsorParty.SubClassOf(parties.PartyInRole);
            
            LenderParty.Define(parties.Role.HasValue(roleIndividuals.Lender));
            AdvisorParty.Define(parties.Role.HasValue(roleIndividuals.Advisor));
            BorrowerParty.Define(parties.Role.HasValue(roleIndividuals.Borrower));
            SponsorParty.Define(parties.Role.HasValue(roleIndividuals.Sponsor));;

            Sponsor = this.DeclareClass<Sponsor>();
            Sponsor.SubClassOf(SponsorParty);
            Equity = Sponsor.DeclareDataProperty<Sponsor, decimal?>(sponsor => sponsor.Equity);
            Sponsor
                .SubClassOf(Equity.ExactCardinality(1))
                .Annotate(
                    validation.Restriction,
                    0);

            Bank             = legalEntities.LegalEntity.DeclareNamedIndividual("Bank");
            BankParty        = this.DeclareClass("BankParty");
            BankLenderParty  = this.DeclareClass("BankLenderParty");
            BankAdvisorParty = this.DeclareClass("BankAdvisorParty");
            BankParty.Define(parties.Organisation.HasValue(Bank));
            BankLenderParty.Define(LenderParty.Intersect(BankParty));
            BankAdvisorParty.Define(AdvisorParty.Intersect(BankParty));

            BankParty.SubClassOf(parties.PartyInRole);

            KeyCounterpartyRole = this.DeclareClass("KeyCounterpartyRole");
            KeyCounterparty     = this.DeclareClass("KeyCounterparty");
            KeyCounterparty.SubClassOf(parties.PartyInRole);
            KeyCounterparty.Define(new ObjectSomeValuesFrom(parties.Role, KeyCounterpartyRole));

            Debt = this.DeclareClass("Debt");
            Debt.SubClassOf(Deal);
            Debt.SubClassOf(Parties.ExactCardinality(1, BankLenderParty));
            Debt.SubClassOf(Borrowers.MinCardinality(1))
                .Annotate(
                    validation.Restriction,
                    0);

            new DisjointClasses(
                this,
                Deal,
                DealType,
                parties.PartyInRole,
                commonDomainObjects.Classifier);

            var ExclusivityClassifier = this.DeclareClass<ExclusivityClassifier>();
            ExclusivityClassifier.SubClassOf(commonDomainObjects.Classifier);
            Deal.SubClassOf(Classifiers.ExactCardinality(1, ExclusivityClassifier))
                .Annotate(
                    validation.Restriction,
                    0)
                .Annotate(
                    validation.SubPropertyName,
                    "Exclusivity");

            var NotExclusive = ExclusivityClassifier.DeclareNamedIndividual("NotExclusive");
            NotExclusive.Value(commonDomainObjects.Id, ExclusivityClassifierIdentifier.No);
            var Exclusive = ExclusivityClassifier.Intersect(new ObjectOneOf(NotExclusive).Complement());

            var ExclusiveDeal = this.DeclareClass("ExclusiveDeal");
            ExclusiveDeal.SubClassOf(Deal);
            var intermediate = this.DeclareClass("Intermediate");
            intermediate.Define(new ObjectSomeValuesFrom(Classifiers, Exclusive));
            ExclusiveDeal.Define(new ObjectIntersectionOf(intermediate));

            var Exclusivity = this.DeclareClass<Exclusivity>();
            var Date = Exclusivity.DeclareDataProperty<Exclusivity, DateTime?>(exclusivity => exclusivity.Date);
            Date.Range(ReservedVocabulary.DateTime)
                .Annotate(
                    validation.RangeValidated,
                    null);
        }
    }

    public interface IDealOntology: IOntology
    {
        IClass Deal { get; }
    }

    public class Advisory:
        Ontology.Ontology,
        IDealOntology
    {
        public IClass Deal { get; protected set; }

        public Advisory(
            CommonDomainObjects commonDomainObjects,
            Validation          validation,
            Deals               deals
            ) : base(
                "Advisory",
                commonDomainObjects,
                validation,
                deals)
        {
            var dealType = deals.DealType.DeclareNamedIndividual("Advisory");
            dealType.Value(
                commonDomainObjects.Id,
                DealTypeIdentifier.Advisory);

            Deal = this.DeclareClass("Deal");
            Deal.SubClassOf(deals.Deal);
            Deal.SubClassOf(deals.Parties.ExactCardinality(1, deals.BankAdvisorParty));
            Deal.SubClassOf(deals.Sponsors.MaxCardinality(0))
                .Annotate(
                    validation.Restriction,
                    0);
        }
    }

    public class ProjectFinance:
        Ontology.Ontology,
        IDealOntology
    {
        public IClass Deal { get; protected set; }

        public ProjectFinance(
            CommonDomainObjects commonDomainObjects,
            Deals               deals,
            Validation          validation
            ): base(
                "ProjectFinance",
                commonDomainObjects,
                deals,
                validation)
        {
            var dealType = deals.DealType.DeclareNamedIndividual("ProjectFinance");
            dealType.Value(
                commonDomainObjects.Id,
                DealTypeIdentifier.ProjectFinance);

            Deal = this.DeclareClass("Deal");
            Deal.SubClassOf(deals.Debt);
            Deal.SubClassOf(deals.Type.HasValue(dealType));
            Deal.SubClassOf(deals.Sponsors.MinCardinality(1))
                .Annotate(
                    validation.Restriction,
                    0);
        }
    }

    public static class ClassificationsExtensions
    {
        public static IDictionary<object, ILookup<IPropertyExpression, ISubClassOf>> Validate(
            this IOntology                    ontology,
            IDictionary<object, ISet<IClass>> classifications
            )
        {
            var evaluator = new ClassMembershipEvaluator(
                ontology,
                classifications);

            return (
                from keyValuePair in classifications
                let individual = keyValuePair.Key
                from classExpression in keyValuePair.Value
                from subClassOf in ontology.Get<ISubClassOf>()
                from annotation in subClassOf.Annotations
                where
                    subClassOf.SubClassExpression == classExpression &&
                    subClassOf.SuperClassExpression is IPropertyRestriction &&
                    annotation.Property.Iri == "Validation.Restriction" &&
                    !subClassOf.SuperClassExpression.Evaluate(
                        evaluator,
                        individual)
                group subClassOf by individual into errorsGroupedByIndividual
                select errorsGroupedByIndividual
            ).ToDictionary(
                group => group.Key,
                group => group.ToLookup(
                    g => ((IPropertyRestriction)g.SuperClassExpression).PropertyExpression));
        }
    }
}
